### PR TITLE
Fix unixmode to handle setuid/sticky when not executable

### PR DESCRIPTION
**Pull request by JamesWTruher**
# PR Summary

This is a fix for #20278 

## PR Context

I'm not addressing the CodeFactor reported complexity as I don't believe this method is overly complicated.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1 [hunk 1]
```diff
@@ -57,9 +61,16 @@ Describe "UnixFileSystem additions" -Tag "CI" {
...
+            # specifically, when setting setgid chmod can fail if the group is privileged.
             chmod "$Mode" "${Item}"
-            $i = Get-Item $Item
-            $i.UnixMode | Should -Be $Perm
+            if ($LASTEXITCODE -ne 0) {
+                set-itresult -skip -because "chmod '$mode' failed"
+            }
+            else {
+                $i = Get-Item $Item
+                $i.UnixMode | Should -Be $Perm
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1:72]:
```suggestion
                $i.UnixMode | Should -BeExactly $Perm
```

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs [hunk 2]
```diff
@@ -594,81 +594,47 @@ public class CommonStat
...
-                            }
-                        }
-                        else
-                        {
-                            modeCharacters[offset] = '-';
-                        }
-
-                        offset++;
-                    }
+                    modeCharacters[0] = itemTypeTable[ItemType];
```
Comment (iSazonov) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:620]: nit:
```c#
                    Span<char> modeCharacters = stackalloc char[10];
```

Also there are two common strings we could return as static (as we do in Windows case):
```
drwxr-xr-x
-rw-r--r--
```

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs [hunk 3]
```diff
@@ -594,81 +594,66 @@ public class CommonStat
...
                 /// <returns>The mode converted into a Unix style string similar to the output of ls.</returns>
                 public string GetModeString()
                 {
-                    int offset = 0;
-                    char[] modeCharacters = new char[10];
-                    modeCharacters[offset++] = itemTypeTable[ItemType];
-
-                    foreach (StatMask permission in permissions)
+                    // On an Ubuntu system (docker), these 3 are roughly 70% of all the permissions
+                    if ((Mode & 0xFFF) == 292)
```
Comment (iSazonov) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:625]: Maybe it is more readable use `UnixFileMode` consts here too instead of `292`?

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs [hunk 4]
```diff
@@ -594,81 +594,66 @@ public class CommonStat
...
-                        }
-
-                        offset++;
+                        return OwnerReadGroupReadOtherRead;
                     }
+                    else if ((Mode & 0xFFF) == 420)
+                    {
+                       return OwnerReadWriteGroupReadOtherRead;
+                    }
+                    else if ((ItemType == ItemType.Directory) & ((Mode & 0xFFF) == 493))
```
Comment (daxian-dbw) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:633]:
```suggestion
                    else if (ItemType == ItemType.Directory & (Mode & 0xFFF) == 493)
```

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs [hunk 5]
```diff
@@ -594,81 +594,66 @@ public class CommonStat
...
-                    StatMask.OtherExecute
-                };
+                private const char NoPerm = '-';
+                private const char SetAndExec = 's';
+                private const char SetAndNotExec = 'S';
+                private const char StickyAndExec = 't';
+                private const char StickyAndNotExec = 'T';
 
                 // The item type and the character representation for the first element in the stat string
                 private readonly Dictionary<ItemType, char> itemTypeTable = new()
```
Comment (daxian-dbw) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:604]: It's better to make this a static member, so you don't create the dictionary every time when creating a `CommonStat` object in `CopyStatStruct`.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs [hunk 6]
```diff
@@ -594,81 +594,66 @@ public class CommonStat
...
-                        }
-                        else
-                        {
-                            modeCharacters[offset] = '-';
-                        }
-
-                        offset++;
+                        return OwnerReadGroupReadOtherRead;
                     }
+                    else if ((Mode & 0xFFF) == 420)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:629]:
```suggestion

                    if ((Mode & 0xFFF) == 420)
```
No need the `else`, given that the previous if block will return.

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Pull request merged**